File: Medium/0951-Flip Equivalent Binary Trees/Recursion.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root1
# @param {TreeNode} root2
# @return {Boolean}
def flip_equiv(root1, root2)
    return true if root1.nil? and root2.nil?
    return false if root1.nil? or root2.nil? or root1.val != root2.val
    return true if flip_equiv(root1.left, root2.left) and flip_equiv(root1.right, root2.right)
    return true if flip_equiv(root1.left, root2.right) and flip_equiv(root1.right, root2.left)
    return false
end

File: Medium/1008-Construct Binary Search Tree from Preorder Traversal/Recursion.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> TreeNode:
        if not preorder:
            return None

        return TreeNode(
            preorder[0],
            self.bstFromPreorder(
                [val for val in preorder if val < preorder[0]]),
            self.bstFromPreorder(
                [val for val in preorder if val > preorder[0]])
        )
